### ISSUE TITLE
Support use of clang::lifetime_capture_by in constructors

### ISSUE THREAD

**Issue by usx95** [llvm/llvm-project#117680]
https://godbolt.org/z/KqM3znMhT
```cpp
struct S{
    S(const int& s [[clang::lifetimebound]]);
};
struct T{
    T(const int& t [[clang::lifetime_capture_by(this)]]);
};

void foo() {
    S s(1); // Error. Good.
    T t(1); // No diagnostic here. Bad.
}
```

**Comment by hokein** [comment]
This case is already supported. The warning is not shown because this warning is disabled by default, see https://godbolt.org/z/rqKhW9qEW.

**Comment by usx95** [comment]
Hmm. I expected it to not work.
I think the current code path should not be taken for constructors and we should use the initialisation path. For example, we have a false positive:
https://godbolt.org/z/rzzM16dGc
```cpp
struct T{
    T(const int& t [[clang::lifetime_capture_by(this)]]);
};

int foo(const T& t);

void foo() {
    T t(1);
    auto x = foo(T(1)); // Error. False positive.
}
```

**Comment by hokein** [comment]
I see. 

I think this is a case (I hit it in #117690 as well) where both the container and temporary object are destroyed at the end of the full expression, and the `lifetimebound` doesn't trigger on this case.


Do we want to diagnose on the following case? I think probably not (in order to reduce noise)

```
void foo() {
    foo(T(1)); 
}
```

**Comment by usx95** [comment]
```cpp
struct T{
    T(const int& t [[clang::lifetime_capture_by(this)]]);
};

int foo(const T& t);
int bar(const T& t [[clang::lifetimebound]]);

void foo() {
    auto x = foo(T(1)); // Should be ok.
    auto y = bar(T(1)); // error is expected.
}
```
I would lean towards treating this as we treat lifetimebound for ctors here: https://github.com/llvm/llvm-project/blob/4a7b56e6e7dd0f83c379ad06b6e81450bc691ba6/clang/lib/Sema/CheckExprLifetime.cpp#L624
And not trigger `checkCaptureByLifetime` for constructors
